packages/server-wallet/src/onchain-service/transaction-submission.ts

```diff
@@ -0,0 +1,120 @@
+import {providers, Wallet} from 'ethers';
+import _ from 'lodash';
+import PriorityQueue from 'p-queue';
+
+import {
+  TransactionSubmissionServiceInterface,
+  MinimalTransaction,
+  TransactionSubmissionOptions,
+} from './types';
+
+// FIXME: replace with
+// import {WalletError as ChannelWalletError} from '@statechannels/server-wallet';
+import {WalletError as ChannelWalletError} from '..';
+
+type Values<E> = E[keyof E];
+
+export class TransactionSubmissionError extends ChannelWalletError {
+  readonly type = ChannelWalletError.errors.OnchainError;
+
+  // Errors where transactions will be safely retried
+  static readonly knownErrors = {
+    badNonce: `the tx doesn't have the correct nonce`,
+    invalidNonce: 'Invalid nonce',
+    noHash: 'no transaction hash found in tx response',
+    underpricedReplacement: 'replacement transaction underpriced',
+  };
+
+  // Reasons an error is thrown from the transaction submission
+  // service
+  static readonly reasons = {
+    zeroAttempts: 'Invalid max transaction submission attempt count of 0',
+    failedAllAttempts: 'Failed all transaction attempts',
+    unknownError: 'Transaction failed with unkown error',
+  } as const;
+  constructor(
+    reason: Values<typeof TransactionSubmissionError.reasons>,
+    public readonly data: any = undefined
+  ) {
+    super(reason);
+  }
+}
+
+export class TransactionSubmissionService implements TransactionSubmissionServiceInterface {
+  // Chain provider
+  private provider: providers.JsonRpcProvider;
+
+  // Wallet to send transactions
+  private wallet: Wallet;
+
+  // Queues to handle transactions
+  private queue = new PriorityQueue({concurrency: 1});
+
+  // Memory tracking wallet nonce to assist with retries
+  // TODO: maybe replace with NonceManager?
```

Comment:
See [`NonceManager`](https://docs.ethers.io/v5/api/experimental/#experimental-noncemanager). I think you would replace the `wallet` private property on this class with that

Reply:
Have you used it before? A little spooked by the fact it's in `experimental` packages

Reply:
We have used the `NonceManager` before for an older state channels hub, and, from what I remember, it worked well!

Reply:
I'm going to leave it as is until it's moved out of `experimental` or until the packages import the `@ethersproject` dependencies.

Just came out of a bit of a nasty ethers upgrade and the versioning was mentioned in the documentation for the experimental packages. Open to using this for v1 if people don't think that will be a problem though!